File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.ts
function numOfMinutes(n: number, headID: number, manager: number[], informTime: number[]): number {
    let g = buildGraph(manager);
    return dfs(headID, g, informTime);
};

function dfs(curr_id, graph: Record<number, number[]>,informTime: number[]): number {
    if(!graph[curr_id] ){
        return 0;
    }           
    let max = 0;
    let sub = graph[curr_id];
    for(let i = 0; i < sub.length; i++){
        max = Math.max(max, dfs(sub[i], graph, informTime));
    }
    
    return max + informTime[curr_id];
}

function buildGraph(manager: number[]): Record<number, number[]> {
    let graph: Record<number, number[]> = {};  
    manager.map((num, i) => {
        if(num != -1){
            if(!graph[num]) graph[num] = [];
            graph[num].push(i);
        }
    })
    return graph;
}
                

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.rs
use std::collections::HashMap;

impl Solution {
    pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
        let graph = Self::build_graph(&manager);
        
        Self::dfs(&head_id, &graph, &inform_time)
    }
    
    fn dfs(
        curr_id: &i32,
        graph: &HashMap<i32, Vec<i32>>,
        inform_time: &Vec<i32>
    ) -> i32 {   
        let mut max_mins = 0;
        
        if !graph.contains_key(curr_id){
            return max_mins;
        }
        for sub in &graph[curr_id] {
            max_mins = max_mins.max(
                inform_time[*curr_id as usize] +
                Self::dfs(
                sub,
                graph,
                inform_time
            ));
        }
        
        return max_mins;
    }
    
    fn build_graph(manager: &Vec<i32>) -> HashMap<i32, Vec<i32>> {
        let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
        for (key, value) in manager.iter().enumerate() {
            if *value != -1 {
                graph.entry(*value).or_insert(Vec::new()).push(key as i32);
            }
        }
        graph
    }
}

//2,2,4,6,-1,4,4,5 -> manager array
//0,1,2,3, 4,5,6,7 -> employees
